### ISSUE TITLE
t.w.xmlrpc.QueryProtocol erroneous newlines in base64 encoding of long credentials

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2980]
|<img alt="stockst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| stockst reported|
|-|-|
|Trac ID|trac#2980|
|Type|defect|
|Created|2008-01-07 02:07:27Z|

In twisted.web.xmlrpc.QueryProtocol.connectionMade(), long Basic Auth credentials (username and password) that result in a base64 encoding > 76 characters have embedded newlines. This results in rejected XML-RPC requests as the Basic Auth RFC (RFC2617) does not impose a 76 character limit and the remote server interprets the newlines as part of the credential. 

This is Python's fault as the string method .encode("base64") is embedding the newlines to break the lines at 76-character boundaries. 

```
>>> "this-is-a-really-long-username:this-is-a-really-long-password".encode('base64')
'dGhpcy1pcy1hLXJlYWxseS1sb25nLXVzZXJuYW1lOnRoaXMtaXMtYS1yZWFsbHktbG9uZy1wYXNz\nd29yZA==\n'
```

The fix is easy, use python's base64 module to do the encoding:

```python
import base64 

def connectionMade(self):
    self.sendCommand('POST', self.factory.path)
    self.sendHeader('User-Agent', 'Twisted/XMLRPClib')
    self.sendHeader('Host', self.factory.host)
    self.sendHeader('Content-type', 'text/xml')
    self.sendHeader('Content-length', str(len(self.factory.payload)))
    if self.factory.user:
        auth = '%s:%s' % (self.factory.user, self.factory.password)
        #auth = auth.encode('base64').strip()
        auth = base64.b64encode(auth.encode)
        self.sendHeader('Authorization', 'Basic %s' % (auth,))
    self.endHeaders()
    self.transport.write(self.factory.payload)
```

Attachments:

* [2980.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b5/7b5d15db9f3aa7bcb19ec97278d1b41d9e041b75/910f856c3ff1dacc96e7ca84d405f825067254d0.diff) (1488 bytes) - added by mwh on 2013-04-11 03:33:46Z - 
* [2980-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b5/7b5d15db9f3aa7bcb19ec97278d1b41d9e041b75/ce48b290ed9f510e5d38fbc723886f1d0ff59dc9.diff) (1835 bytes) - added by mwh on 2013-04-11 03:38:31Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2980 2980
type__defect defect
reporter__stockst stockst
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1199671647000000 1199671647000000
changetime__1365693339000000 1365693339000000
version__None None
owner__therve therve
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="stockst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|stockst commented|
|-|-|

Replying to [stockst](#2980):

Grr, typo, the fix should be:

```python
import base64
def connectionMade(self):
    self.sendCommand('POST', self.factory.path)
    self.sendHeader('User-Agent', 'Twisted/XMLRPClib')
    self.sendHeader('Host', self.factory.host)
    self.sendHeader('Content-type', 'text/xml')
    self.sendHeader('Content-length', str(len(self.factory.payload)))
    if self.factory.user:
        auth = '%s:%s' % (self.factory.user, self.factory.password)
        #auth = auth.encode('base64').strip()
        auth = base64.b64encode(auth)
        self.sendHeader('Authorization', 'Basic %s' % (auth,))
    self.endHeaders()
    self.transport.write(self.factory.payload)
```

Sorry about that.

**Comment by twisted-trac** [comment]
|<img alt="stockst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|stockst commented|
|-|-|

Not fixed in Twisted 8.x, bump.

Replying to [stockst](#2980):
> In twisted.web.xmlrpc.QueryProtocol.connectionMade(), long Basic Auth credentials (username and password) that result in a base64 encoding > 76 characters have embedded newlines. This results in rejected XML-RPC requests as the Basic Auth RFC (RFC2617) does not impose a 76 character limit and the remote server interprets the newlines as part of the credential.
>
> This is Python's fault as the string method .encode("base64") is embedding the newlines to break the lines at 76-character boundaries.
>
> ...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Could you write a unit test that fails without the fix?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_5"></a>|@mwhudson commented|
|-|-|

Writing a unit test that fails is easy, see [2980-2.diff.](None/2980-2.diff.)  The way it fails without the code change isn't especially clear due to Twisted being stricter at interpreting newlines in headers than RFC 2616 suggests and more tolerant than I would expect of bogus authorization headers.  But it's probably OK?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

Looks good, applying.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [38003]) Use the base64 module in xmlrpc QueryProtocol to generate valid authorization
header for long user names and passwords.

Author: mwh
Reviewer: therve
Fixes: [#2980](https://github.com/twisted/twisted/issues/2980)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set status to `reopened`|
|-|-|

The topfile was placed in the `words` project where it should be in `web`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

The correct way to re-open a ticket is by reverting the revision that closed it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `reopened`|
|-|-|

(In [38004]) Revert r38003: NEWS fragment in the wrong place

Reopens: [#2980](https://github.com/twisted/twisted/issues/2980)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [38005]) Use the base64 module in xmlrpc QueryProtocol to generate valid authorization
header for long user names and passwords. This is a re-merge of r38003 with
NEWS fragment in the right place.

Author: mwh
Reviewer: therve
Fixes: [#2980](https://github.com/twisted/twisted/issues/2980)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Thanks!
